### PR TITLE
[precompile] Compose a training graph to standalone source

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.17.0) (oldest at bottom):
* #195352
* #195351
* #195664
* #195350
* __->__ #195349
* #195348
* #195347
* #195346

A dynamo capture is worth serving only if the artifact carries everything the
capture compiled. It did not: a training capture shipped as an opaque pickled
bundle instead of readable source, and an eager capture silently lost graph
nodes on the way back in.

aot_autograd.compile_to_python renders a graph standalone by splicing
AOTAutograd's own codegen'd wrapper source around an inner Inductor call, and
did that only for inference; a joint reached the inner compiler twice and was
refused as "more than one inner forward graph". AOTAutograd already codegens
the backward's bodies -- compiled_function_forward, compiled_function_backward
and the backward prologue and epilogue beside them -- and the
torch.autograd.Function holding them is ordinary Python, so the composer emits
it. It also needs AOTAutograd's spec (fw_metadata, RNG state, disable_amp, the
saved and symint counts), which is built during the capture pass and is not
reachable from outside it, so AOTDispatchAutograd.post_compile gains a
thread-local sink, capture_autograd_compile_specs(), that records every spec
built while active; the composer selects the one whose compiled forward is
the graph it captured and refuses when zero or several match. The context is
nestable; on exit it removes its own sink by identity, scanning from the most
recent, so generator-driven contexts that exit out of order neither lose specs
nor raise from the finally block. compile_to_python gains grad_enabled: the
capture pass runs under no_grad by default, which pins the inference path, and
under enable_grad AOTAutograd emits the joint. rendered_backends passes the
session's training flag through it, so the up-front refusal that kept a
training capture pickled is lifted and keep_graphs no longer excludes
training. The two spliced Inductor modules are renamed per module by
namespace_module_names, for the same reason two shape variants of one frame
are: each defines call, Runner and its kernels and resolves them as late-bound
globals. The RNG-state tracker is emitted as a constructor call over the three
fields AOTAutograd builds it from rather than as a baked instance: every other
field is per-run state its default factory rebuilds, and one of them defaults
to an itertools.count, whose pickle support CPython removed in 3.14. Baking it
there leaves the emitter with no reduce and the whole compose is rejected,
which a caller only sees as an artifact that silently fell back to a pickled
bundle. Composition refuses up front rather than mis-composing: a graph with
tensor subclasses, one carrying a BackwardState, one where AOTAutograd did not
codegen the five wrapper sources expected, and one carrying any other active
runtime wrapper -- functionalized RNG, dedup, synthetic bases, effect tokens,
debug asserts, an aliasing output or an out-of-graph input mutation -- all
raise NotImplementedError. One limitation is not a compose-time refusal and
shows up at serve time instead: the emitted autograd.Function sets
boxed_grads_call and carries no _aot_id, so running a served artifact's
backward under compiled autograd raises from compiled_autograd.py rather than
tracing. The artifact is self-contained by design and never joins an AOT
backward, so this is a real boundary rather than a bug, and the two tests that
would cross it opt out of compiled autograd explicitly.

The ordering carries the risk. Layout optimization lets a compiled forward
hand back channels-last saved activations, so the backward must be lowered
against the strides the forward ACTUALLY chose. torch.compile takes those from
TracingContext.report_output_strides and applies them in
_aot_stage2b_bw_compile; a capture pass that never lowers cannot observe them,
and two graphs lowered separately afterwards then disagree -- a loud
AssertionError on a conv net, silently wrong gradients with size asserts off.
Inductor's compile_to_python had the strides and discarded them; it now
reports them out as the stride expressions Inductor prints, and the composer
lowers the forward, rebuilds those expressions in the graph's ShapeEnv,
compares them under suppress_guards the way _aot_stage2b_bw_compile does,
restrides the backward's placeholders (symbolically where the shapes are),
then lowers the backward with is_backward=True, as compile_fx does. The
restride lands on node.meta["val"] rather than an example-inputs list, since
that entry point rebuilds its fakes from placeholder metadata and ignores the
inputs it is handed, so restriding a list would be a silent no-op.

That same inner lowering call was hardcoded is_inference=True. compile_fx
already passes the right value to inner_compile for every graph it lowers, so
the capture records is_inference from those kwargs per graph and hands it back
to Inductor rather than guessing from op names. A make_fx graph that
differentiates inline (tracing through .backward()) has no output requiring
grad, so AOTAutograd dispatches it down the inference path and compile_fx
lowers it with is_inference=True; recording the decision keeps the composed
lowering in agreement with torch.compile's for the same graph.

A training example runs a real backward, which ACCUMULATES into the caller's
tensors, so a capture left .grad changed on the model it was handed and, on
the documented warmup-step-then-capture flow, doubled it. The dynamo session
now snapshots every tensor an example could accumulate into -- the entry's own
parameters and buffers, and every tensor or module among the examples -- clears
.grad before the example calls and restores the SAME grad objects in the
finally, as the make_fx path already did in _capture: a caller holding a prior
p.grad reference, or optimizer state keyed on grad identity, must not be
invalidated by having been used as an example.

An eager subgraph ships as a pickled GraphModule, whose reduction deliberately
keeps only the generated source and recovers the Graph by re-tracing it. For a
Dynamo graph that is lossy: a HOP rejects the Proxy and load raises, and,
worse because silent, an in-graph context manager takes no Proxy at all, so
the retrace EXECUTES it and records nothing -- serving fp32 where capture
chose bf16, and leaving grad mode flipped in load()'s caller. The top level of
a Dynamo graph is only ever CALLED and nothing reads its .graph, so
_SourceGraphModule rebuilds the module from that same source and carries no
Graph at all. Nested HOP bodies do need one, since some eager HOP
implementations run them through fx.Interpreter, so those alone go through
GraphPickler. _SourceGraphModule defines __deepcopy__ for the same reason
GraphModule does: without it deepcopy falls through to __reduce__, which
re-execs the source and re-loads every nested body on every copy, and
PrecompileContext.record_artifact copies under no_dispatch(), where rebuilding
a fake tensor asserts.

None of the guard-serialization work is gated on precompile. It sits in
CheckFunctionManager.serialize_guards and pickle_guards_state, so it also
changes what torch._dynamo.aot_compile and caching_precompile write, and
owners there should look. The largest piece is one defect seen twice. Pruning
unreached values to a _Missing sentinel was too narrow, running only for
nn.Module, so a guarded object that is not a module -- a pipeline object
holding a dataloader -- was pickled whole and one unguarded attribute, a live
generator, took the whole frame down with "cannot pickle 'generator' object",
which surfaced only as the entry frame producing no guarded code. It was also
too blunt, keying on id(), which asks "same OBJECT" where it means "same
REFERENCE": an unguarded attribute holding torch.float32 marks that one
interned object missing, so every other reference to that dtype resolves to
the sentinel and the artifact fails at LOAD with "argument 'dtype' must be
torch.dtype, not _Missing". Pruning now reaches any guarded user object and
skips interned values, which it gained nothing from anyway. It also skips any
user object that decides its own reduction -- one with __reduce__,
__getstate__, __setstate__ or __getnewargs__, or a reducer registered for its
class in copyreg.dispatch_table, which sits outside the class and so shows up
on no hook -- because the sentinel pruning leaves behind would arrive as a
constructor argument or a field its __setstate__ reads and reconstruct the
object wrong, rather than merely losing an attribute nobody guards. Each hook
is compared against object's own slot by identity, since object grew a default
__getstate__ in 3.11, and the __getnewargs__ a tuple or a
collections.namedtuple carries is treated as benign: it returns the items,
which live outside __dict__ and are never pruned, so a namedtuple subclass
with extras is still pruned.

Re-serializing guard state, which the invariant-guard policy does from the
pickle the capture already made, runs the reducers a second time, and that
exposed one that read live state where it needed the recorded one. The
bound-method reducer emitted a bound method and then decided, on the next
pass, whether to take that branch again by READING the receiver, which pruning
had by then replaced with the sentinel, aborting capture with "'_Missing'
object has no attribute 'forward'"; it decides from the receiver's fate
instead, so the branch is a fixed point over its own output.
test_precompile_serves_a_bound_method_whose_receiver_was_pruned pins it over
the torchrec pipeline shape that hit it -- a saved bound forward whose receiver
is pruned, with and without the attribute rebound, read and unread -- and
asserts the artifact SERVES, since an earlier fix that degraded the method
itself to the sentinel captured fine and then never matched a live receiver.

Re-serializing a pickle needs two more fixes: the tensor reducer read live
state where it wanted the recorded snapshot, which by then is the fake the
first pass wrote, so it reported dispatch keys the real tensor never had and
dragged FakeTensorMode into the artifact; and serialize_guards indexed the
builtins dict unguarded, which pickle_guards_state drops from an all-portable
capture while keeping its name. Guard.code_list and Guard.guard_types stop
being pickled, since create_fn repopulates them at load while set_export_info
EXTENDS them on every build. A concrete process group joins
_get_unsupported_types: it is bound as a subclass of Backend rather than of
ProcessGroup, so the existing entry missed it and a FakeProcessGroup killed
the frame.

Two diagnostics answer the same complaint, that a failure named a type and
never a place. A pickle failure now reports the attribute path to the
offending value, found by IDENTITY: GuardsStatePickler records the object
reducer_override was last handed, and _offending_value_path walks the guard
state for that object rather than for anything of the failing type, which
would name a same-typed bystander. The walk starts where the pickler starts,
at the state itself, with the two scopes as seeds ahead of it in the queue, so
a value on a compiler internal reachable only through the guards is still
found while the common case still reports the short scope path; and the scope
roots are snapshotted before pickle_guards_state empties global_scope for an
all-portable capture, which is how a real "cannot pickle '_thread.RLock'
object" arrived with no path at all. The same path is appended when a reducer
itself refuses a value. And a bypassed frame records its reason, so "captured
no dispatchable graph" says that rather than blaming a thin wrapper that was
never the problem.

A compiled subgraph is Inductor output and has a source form, so one that
fails to render is not dropped silently: rendered_backends warns once and
records it in PrecompileSummary.unrendered_backends, which the header carries
as UNRENDERED_BACKENDS and the header parser reads back as an optional
section, so a training artifact that quietly fell back to the pickled bundle
is visible rather than looking complete.

Read it following a captured graph: the AOTAutograd composer, the session's
rendering and gradient handling, the eager-backend round trip, then the
guard-serialization fixes underneath all of it.

Test Plan:

```
python test/test_precompile.py
python test/functorch/test_compile_to_python.py
python test/dynamo/test_package.py
python test/dynamo/test_guard_serialization.py
```

In test_compile_to_python.py, test_training_graph_composes_forward_and_backward
checks that a served training output carries grad_fn and matches eager's
gradients, test_training_forward_and_backward_do_not_share_names that the two
spliced Inductor modules share no top-level name, which is what stops the
forward resolving the backward's kernels as late-bound globals and calling them
with the wrong arity, test_training_dynamic_saved_activation_stride_runs_like_eager
and test_restride_backward_placeholders_keeps_symbolic_strides pin the
symbolic restride on a saved activation whose stride is a shape expression
rather than an int, the three test_training_rejects_* tests pin the compose-time
refusals, and test_nested_autograd_spec_sinks_exit_by_identity and
test_interleaved_autograd_spec_sink_exits the sink's exit order. In
test_precompile.py, the test_eager_backend_* round trip is a matrix over cond,
while_loop, checkpoint, vmap and autocast, each with and without a graph break
so both serving modes are covered, checked against
torch.compile(backend="eager"); test_capture_does_not_touch_the_example_gradients
takes a warmup step and then asserts, over both tracers, that each parameter's
grad is the same object with the same value after the capture;
test_unpicklable_value_error_names_the_attribute_path,
test_unpicklable_value_reachable_only_through_a_global_is_named,
test_unpicklable_value_outside_both_scopes_is_named and
test_unpicklable_guard_value_names_where_it_lives pin the identity walk, its
seeds and the snapshotted roots;
test_serialized_guards_drop_export_bookkeeping,
test_reserialized_guards_do_not_carry_the_fake_tensor_machinery,
test_guarded_user_object_prunes_its_unguarded_attributes and
test_unguarded_interned_attribute_does_not_poison_the_artifact pin the
serialization fixes; the two test_no_dispatchable_graph_* tests the bypass
reason; test_unrendered_subgraphs_are_reported_and_listed_in_the_header that
a subgraph whose compose fails is reported in the summary and the header rather
than dropped; and test_dynamo_tracer_renders_kernels_as_source now expects
readable source for its training case too, where it used to accept the opaque
bundle. In test_guard_serialization.py,
test_guard_on_an_object_that_reconstructs_itself,
test_guard_on_an_object_whose_setstate_or_new_reads_state,
test_guard_on_a_copyreg_registered_class and
test_guard_on_a_namedtuple_subclass_with_an_unpicklable_extra pin the
pickle-protocol rule for attribute pruning. In test_package.py,
test_cache_entry_loads_from_a_pickle_without_newer_fields now covers the
bypass_reason field this commit adds.

Authored with the assistance of an AI coding agent.

### CODE REVIEW

**Comment by bobrenjc93** [comment]
Superseded by #195361/#195362 in the canonical precompile stack #195353-#195362. Not carried over: the `_builds_its_own_pickle` rule (keep objects with their own __reduce__/copyreg whole) and its four tests; worth porting as a follow-up. Closing.
